content/en/blog/_posts/2020-12-10-Granular-Control-of-Volume-Permission-Changes.md

```diff
@@ -0,0 +1,58 @@
+
+When configuring a pod’s security context, configure fsGroupChangePolicy with “OnRootMismatch” so if the root of the volume already has the correct permissions, the recursive permission change can be skipped. Kubernetes ensures that permissions of the top-level directory are changed last the first time it applies permissions. 
+
+```yaml
+securityContext:
+  runAsUser: 1000
+  runAsGroup: 3000
+  fsGroup: 2000
+  fsGroupChangePolicy: "OnRootMismatch"
+```
+More details about this can be found [here](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods).
+
+### Allow CSI Drivers to declare support for fsGroup based permissions
+
+While the previous section implied that we ALWAYS recursively change permissions of a volume if Pod has a fsGroup, this is not strictly true. For certain multi-writer volume types, such as NFS or Gluster, we don’t perform recursive permission changes even if the pod has a fsGroup. Other volume types may not even support chown/chmod which are unix-based permission control primitives. 
+
+So how do we know when to apply recursive permission changes and when we shouldn’t? For in-tree storage drivers, this was relatively simple. But for CSI drivers that could span a multitude of platforms and storage types, this problem can be challenging. 
+
+Previously, whenever a CSI volume was mounted to a Pod, Kubernetes would attempt to automatically determine if the permissions and ownership should be modified. These methods were imprecise and could cause issues, outlined previously, depending on the storage type.
+
+The FSGroupPolicy field introduced in the CSIDriver spec allows drivers to explicitly opt in or out of these recursive modifications. By having the CSI driver specify a policy for the backing volumes we can avoid needless modification attempts, reducing mount time and avoiding errors.
+
+#### CSIDriver FSGroupPolicy API
+
+Three FSGroupPolicy values are available as of Kubernetes 1.20, with more planned for future releases.
+
+- ReadWriteOnceWithFSType - This is the default policy, applied if no FSGroupPolicy is defined, and preserves the behavior from previous Kubernetes releases. Each volume is examined at mount time to determine if permissions should be recursively applied.
+- File - Always attempt to apply permission modifications, regardless of the filesystem type or PersistentVolumeClaim’s access mode.
+- None - Never apply permission modifications.
+
+#### How do I use it?
+The only configuration needed is defining an FSGroupPolicy inside of the CSIDriver’s spec. Once that element is defined, any subsequently mounted volumes will automatically use the defined policy. There’s no additional deployment required!
```

Comment:
You should mention that it's important to check if the driver can support this.

Reply:
@msau42 To clarify, by "this" do you mean the FSGroupPolicy API?  Thanks!

Reply:
I mean support this functionality of changing volume ownership via chmod/chown. 

I mainly don't want to imply that any Kubernetes user can just set this field and expect it to work. It requires the csi driver to be able to support it.

cc @gnufied

Reply:
Ok, I'll wait for @gnufied to suggest some additional language here.  Thank you @msau42 .

Reply:
So I think most k8s admins will not have to set this manually and a CSI driver deployment should configure this automatically. But for completeness sake we can say:

"If you are a CSI driver author or you are deploying a CSI driver, ensure that driver deployment scripts install `CSIDriver` object with appropriate value of `FSGroupPolicy` that is supported by the driver`.